### PR TITLE
Add `<iotune>` support for DataVolume and PVC source

### PR DESCRIPTION
**Pull request by zulh-civo**
**What this PR does / why we need it**:

To close https://github.com/kubevirt/kubevirt/issues/5444 by adding `<iotune>` to Libvirt XML with these 6 key fields ([source](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/virtualization_tuning_and_optimization_guide/sect-virtualization_tuning_optimization_guide-blockio-techniques)):

* `total-bytes-sec`
    The total throughput limit in bytes per second. 
* `read-bytes-sec`
    The read throughput limit in bytes per second. 
* `write-bytes-sec`
    The write throughput limit in bytes per second. 
* `total-iops-sec`
    The total I/O operations limit per second. 
* `read-iops-sec`
    The read I/O operations limit per second. 
* `write-iops-sec`
    The write I/O operations limit per second. 

_YAML example 1_

```yaml
apiVersion: kubevirt.io/v1alpha3
kind: VirtualMachine
metadata:
  labels:
    kubevirt.io/vm: testvm1
  name: dv-vm
spec:
  dataVolumeTemplates:
    - metadata:
        name: awesome-dv
        annotations:
          iotune.kubevirt.io/read-bytes-sec: "52428801"
          iotune.kubevirt.io/write-bytes-sec: "52428802"
          iotune.kubevirt.io/read-iops-sec: "52428803"
          iotune.kubevirt.io/write-iops-sec: "52428804"
      spec:
        pvc:
          accessModes:
            - ReadWriteOnce
          resources:
            requests:
              storage: 5Gi
        source:
          http:
            url: https://cloud-images.ubuntu.com/xenial/current/xenial-server-cloudimg-amd64-disk1.img
  running: true
  template:
    metadata:
      labels:
        kubevirt.io/vm: testvm1
    spec:
      domain:
        cpu:
          cores: 1
        devices:
          disks:
            - name: disk0
              disk:
                bus: virtio
            - name: cloudinitdisk
              disk:
                bus: virtio
        resources:
          requests:
            memory: 1Gi
      volumes:
        - dataVolume:
            name: awesome-dv
          name: disk0
        - cloudInitNoCloud:
            userData: |
              #cloud-config
              chpasswd:
                list: |
                  ubuntu:12341234
                expire: false
              hostname: vm
              ssh_pwauth: yes
              disable_root: false
          name: cloudinitdisk
```

_YAML example 2_

```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: ubuntu
  labels:
    app: containerized-data-importer
  annotations:
    iotune.kubevirt.io/total-bytes-sec: "52428800"
    iotune.kubevirt.io/total-iops-sec: "1000"
    cdi.kubevirt.io/storage.import.endpoint: "https://cloud-images.ubuntu.com/xenial/current/xenial-server-cloudimg-amd64-disk1.img"
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 5Gi
---
apiVersion: kubevirt.io/v1alpha3
kind: VirtualMachine
metadata:
  creationTimestamp: 2018-07-04T15:03:08Z
  generation: 1
  labels:
    kubevirt.io/os: linux
  name: vm
spec:
  running: true
  template:
    metadata:
      creationTimestamp: null
      labels:
        kubevirt.io/domain: vm
    spec:
      domain:
        cpu:
          cores: 2
        devices:
          disks:
          - disk:
              bus: virtio
            name: disk0
          - cdrom:
              bus: sata
              readonly: true
            name: cloudinitdisk
        machine:
          type: q35
        resources:
          requests:
            memory: 1024M
      volumes:
      - name: disk0
        persistentVolumeClaim:
          claimName: ubuntu
      - cloudInitNoCloud:
          userData: |
            #cloud-config
            chpasswd:
              list: |
                ubuntu:12341234
              expire: false
            hostname: vm
            ssh_pwauth: yes
            disable_root: false
        name: cloudinitdisk
```

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubevirt/kubevirt/issues/5444

**Special notes for your reviewer**:
None

**Release note**:
```release-note
* Added `<iotune>` support for DataVolume and PVC source
```

### CODE REVIEW

**Comment by zulh-civo** [comment]
TBH, I don't know how to edit all my previous commits in this PR to add the DCO line (e.g. `Signed-off-by: My Name <name@example.org>`).

Plus, I'm afraid it will cause out-of-sync problem to reviewers. If you think it's fine to proceed, could you let me know how to amend all my previous commits to include that DCO line? Thanks.

**Comment by alicefr** [comment]
@zulh-civo you have some print that I think are only for development, you might want to remove them from the PR.  You might also want to remove all the Revert commits. You can squash the commits and sign the commits you want to keep and force push

**Comment by alicefr** [comment]
@zulh-civo you have some print that I think are only for development, you might want to remove them from the PR.  You might also want to remove all the Revert commits. You can squash the commits and sign the commits you want to keep and force push

**Comment by alicefr** [comment]
Can you also describe how you pass the various values to the PVC and DV?

File: pkg/virt-launcher/virtwrap/converter/converter.go [hunk 1]
```diff
@@ -626,86 +640,147 @@ func GetHotplugBlockDeviceVolumePath(volumeName string) string {
...
 	disk.Source.Dev = GetHotplugBlockDeviceVolumePath(volumeName)
+
+	disk.Iotune.TotalBytesSec = iotune.TotalBytesSec
+	disk.Iotune.TotalIopsSec = iotune.TotalIopsSec
+	disk.Iotune.ReadBytesSec = iotune.ReadBytesSec
+	disk.Iotune.WriteBytesSec = iotune.WriteBytesSec
+	disk.Iotune.ReadIopsSec = iotune.ReadIopsSec
+	disk.Iotune.WriteIopsSec = iotune.WriteIopsSec
+
+	// disk.Iotune.TotalBytesSec = 52428800
```
Comment (alicefr) [inline pkg/virt-launcher/virtwrap/converter/converter.go:761]: I guess this part has to be removed too

File: .gitignore [hunk 2]
```diff
@@ -45,3 +45,4 @@ hack/builder/manifests/
 .bazeldnf/
 user.bazelrc
 .vscode/
+tests/*.log
```
Comment (alicefr) [inline .gitignore:48]: this also has to be removed from the PR

**Comment by zulh-civo** [comment]
> Can you also describe how you pass the various values to the PVC and DV?

Sorry, I'm not clear with your question. What do you mean?

File: pkg/virt-launcher/virtwrap/converter/converter.go (hunk 1 above)
Comment (zulh-civo) [reply pkg/virt-launcher/virtwrap/converter/converter.go:761]: Will do

File: .gitignore (hunk 2 above)
Comment (zulh-civo) [reply .gitignore:48]: Will do

**Comment by alicefr** [comment]
Which annotations you use to set those values. It will be nice to have that in the description of the PR. Like `iotune.kubevirt.io/total-bytes-sec` An example will be also very helpful, like a PVC with all those annotations

File: .gitignore (hunk 2 above)
Comment (zulh-civo) [reply .gitignore]: Removed

File: pkg/virt-launcher/virtwrap/converter/converter.go (hunk 1 above)
Comment (zulh-civo) [reply pkg/virt-launcher/virtwrap/converter/converter.go]: Removed

**Comment by zulh-civo** [comment]
@alicefr - thank you for taking time checking this PR for me. Here are the answers for your questions:

> You have some print that I think are only for development, you might want to remove them from the PR.

Removed

> You might also want to remove all the Revert commits.

Removed

> You can squash the commits and sign the commits you want to keep and force push.

Done

> Can you also describe how you pass the various values to the PVC and DV?

By using annotations on  the PVC and DV (please refer to my PR description

> Which annotations you use to set those values. It will be nice to have that in the description of the PR. Like `iotune.kubevirt.io/total-bytes-sec`. An example will be also very helpful, like a PVC with all those annotations.

I've updated the PR description with examples

Please do let me know if you need any extra information. Thank you.

**Comment by alicefr** [comment]
@zulh-civo it looks much better :). A small suggestion from my side, you could send an email to get feedback on your proposal in [kubevirt-dev](https://groups.google.com/g/kubevirt-dev) explaining why you need to set those values

**Comment by zulh-civo** [comment]
@alicefr - thank you for the tip! I've opened a thread [here](https://groups.google.com/g/kubevirt-dev/c/fGRuWW7bPUQ).

**Comment by zulh-civo** [comment]
@ashleyschuett @omeryahud - I really appreciate if you can review this. Thank you.

File: pkg/virt-handler/vm.go [hunk 3]
```diff
@@ -2414,6 +2415,49 @@ func (d *VirtualMachineController) vmUpdateHelperDefault(origVMI *v1.VirtualMach
...
+			} else {
+				name = source.DataVolume.Name
+			}
+
+			pvc, err := d.clientset.CoreV1().PersistentVolumeClaims(vmi.Namespace).Get(context.Background(), name, metav1.GetOptions{})
+			if err != nil {
+				continue
+			}
+
+			annotations := pvc.ObjectMeta.Annotations
```
Comment (mhenriks) [inline pkg/virt-handler/vm.go:2435]: Looks like CDI datavolume controller will have to be updated to pass iotune annotations to the pvc

Comment (zulh-civo) [reply pkg/virt-handler/vm.go:2435]: When I was testing this on my dev machine, I didn't change anything in CDI controller. I applied 2 example YAMLs above and they worked.

FYI, I was looking at [this line](https://github.com/kubevirt/kubevirt/blob/81e8bb646fb3bc3de4b47018acfa5d5e78cdc495/pkg/virt-handler/vm.go#L2411) before I started this PR. It gave me some clue that no update CDI controller is required.

If I misunderstood you, please explain again. Thanks.

**Comment by Icedroid** [comment]
@zulh-civo @mhenriks why close this issus, It is very helpful for us. Cloud be merge to kubevirt?

**Comment by alicefr** [comment]
@Icedroid it got stale and it was automatically closed. Let me try to reopen it

/reopen

**Comment by alicefr** [comment]
However, it seems that we don't want to add these fields into the kubevirt API. The suggested workaround has been to use a sidecar (see https://github.com/kubevirt/kubevirt/issues/5444#issuecomment-834560530)

**Pull request closed without merging**